[CORE-9167] Refactored Raft protocol scheduling groups

mmaslankaprv opened the pull request:
The `consensus::timeout_now` method triggers an immediate leader
election to handle the leadership transfers. Since the method is called
directly from the RPC request handler it is executed in the scheduling
group that the handler uses. In this case it is `raft` scheduling group.
This makes the whole leader election process to run in `raft` scheduling
group while it normally runs in `main`. Changed the code to always
handle timeout now in `main` scheduling group not to introduce
differences after the leadership has been transferred.

Fixes: CORE-9167

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
r/consensus: changed timeout_now method implementation to coroutine

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/consensus: switch back to main scheduling group in timeout_now

The `consensus::timeout_now` method triggers an immediate leader
election to handle the leadership transfers. Since the method is called
directly from the RPC request handler it is executed in the scheduling
group that the handler uses. In this case it is `raft` scheduling group.
This makes the whole leader election process to run in `raft` scheduling
group while it normally runs in `main`. Changed the code to always
handle timeout now in `main` scheduling group not to introduce
differences after the leadership has been transferred.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1ff10472-b105-4c2c-8b2f-a55b87dbf78a-->
<!--TEST RESULTS-->
<!--8e4ad7fe-2537-4db3-b0a5-ca4479658d1a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62261">build#62261</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62261#01953cad-023c-47f4-84df-2400c0eeb0be | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62261#01953ca8-9f26-4860-b1df-c86b77221eb9 | FLAKY | 1/2 |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move.replication_factor=3.unclean_abort=False.recovery=restart_recovery.compacted=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/62261#01953ca8-9f26-4860-b1df-c86b77221eb9 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62261#01953cad-023d-4fab-b659-215c0c14108b | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62392">build#62392</a></summary>


| test_id                                                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62392#01954697-7e36-45dd-be17-a90c4ff1dbb7 | FLAKY         | 1/2      |
</details>
<!--/body-->

mmaslankaprv commented on src/v/raft/consensus.cc:
```diff
@@ -49,6 +49,7 @@
 #include <seastar/core/gate.hh>
```
it is only election and replication of the configuration batch, so this one may drive one `replicate_batcher::flush` but this effect should not be permanent

mmaslankaprv replied:
I checked that once again. CORE-9167 is caused by the fact that the dispatch loop of `raft::buffered_protocol` is created with different scheduling group. The group it uses may be one of: 

- `produce`
- `raft`
- `main`
The group that is going to be used is determined what was the context the first append entries request was called with. If node was idle and queue is created as a downstream of produce request it would be `produce` after leadership transfer it is `raft` as the group configuration is replicated within that group. When topic was created and there is an immediate load then the `main` group will be used as this is the one which we replicate configuration during first leader election.

mmaslankaprv replied:
I will update this PR with another commit but i am wondering which group should the dispatch loop be run in

StephanDollberg replied:
What was the motivation for that?

mmaslankaprv replied:
We were wondering in which group the `dispatch_loop` should be scheduled in.

StephanDollberg replied:
Right but why add another one? I think a while ago we had discussed having everything on the leader in produce and the follower work in raft.

With three produce related groups produce is now weighted 3 times as much as fetch. As always hard to reason about but it feels it just gets more complicated with more groups.

StephanDollberg replied:
> Right, but this was (part of) "the revolution" and it didn't really happen, right?

Yes, though I thought that failed not because of that but because of making the kafka group only 100 shares (op lock batching etc.)

> Produce seems wrong, after all this is generic raft stuff which may be driven from anywhere, not just produce (though of course produce is a major driver in practice).

I wasn't aware there was other stuff than append_entries goes through buffered_protocol as well. I would have put it in produce as I thought that was the main (and only driver).

> Also, having produce weighted more than fetch seems "fine"?

Yeah agree, though you probably also don't want to take more in than you can get out (but I don't think that's a massive issue because of how fetch works). 

> I'm not married to it though: what is your suggestion?

Was mostly just trying to understand the motivation as it seemed yet different to what we had discussed in the past. Because of the usual SG stuff I can't really reason for one way or the other so not saying this is bad or anything.

StephanDollberg replied:
Fair, I guess from an observability pov it's good to have them separate.

mmaslankaprv replied:
Is the conclusion that we should keep what is in this PR ?

mmaslankaprv committed:
cpu_scheduling: renamed raft scheduling group to better reflect its use

Renamed the `raft` scheduling group to `raft_recv`. The group is used
only when Raft protocol receives requests, this way the group name
better reflects it actual usage.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
scheduling: repurpose raft learner recovery scheduling group

Repurposing Raft learner recovery scheduling group to handle Raft send
side. Raft learner recovery is not being dispatched in the special
group hence the group remained not used. The group is going to be used
to schedule raft leader election and buffered protocol dispatch loop in
it.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/types: remove unused learner recovery scheduling primitives

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
io_priority: removed not used learner io_priority class

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
raft: wire in raft send scheduling group

Pass `raft_send` scheduling group into `raft::group_manager` and down to
the `consensus` instances.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
raft: use raft send scheduling group

Using raft send scheduling group throughout the protocol
implementation. The send scheduling group is used for the following:
- dispatch loop in `raft::buffered_protocol`
- leadership and committed offset notifications
- recovery
- leader election

Signed-off-by: Michał Maślanka <michal@redpanda.com>
